util/stmtsummary/evicted.go

```diff
@@ -0,0 +1,158 @@
+package stmtsummary
+
+import (
+	"container/list"
+	"time"
+
+	"github.com/pingcap/parser/mysql"
+	"github.com/pingcap/tidb/types"
+)
+
+// stmtSummaryByDigestEvicted contents digests evicted from stmtSummaryByDigestMap
+type stmtSummaryByDigestEvicted struct {
+	// record evicted data in intervals
+	// latest history data is Back()
+	history *list.List
+}
+
+// element being stored in stmtSummaryByDigestEvicted
+type stmtSummaryByDigestEvictedElement struct {
+	// beginTime is the begin time of current interval
+	beginTime int64
+	// endTime is the end time of current interval
+	endTime int64
+	// *Kinds* of digest being evicted
+	digestKeyMap map[string]struct{}
+}
+
+// spawn a new pointer to stmtSummaryByDigestEvicted
+func newStmtSummaryByDigestEvicted() *stmtSummaryByDigestEvicted {
+	return &stmtSummaryByDigestEvicted{
+		history: list.New(),
+	}
+}
+
+// spawn a new pointer to stmtSummaryByDigestEvictedElement
+func newStmtSummaryByDigestEvictedElement(beginTime int64, endTime int64) *stmtSummaryByDigestEvictedElement {
+	return &stmtSummaryByDigestEvictedElement{
+		beginTime:    beginTime,
+		endTime:      endTime,
+		digestKeyMap: make(map[string]struct{}),
+	}
+}
+
+// AddEvicted is used add an evicted record to stmtSummaryByDigestEvicted
+func (ssbde *stmtSummaryByDigestEvicted) AddEvicted(evictedKey *stmtSummaryByDigestKey, evictedValue *stmtSummaryByDigest, historySize int) {
+	evictedValue.Lock()
+	defer evictedValue.Unlock()
+
+	for e, h := evictedValue.history.Back(), ssbde.history.Back(); e != nil; e = e.Prev() {
+		evictedElement := e.Value.(*stmtSummaryByDigestElement)
+		eBeginTime := evictedElement.beginTime
+		eEndTime := evictedElement.endTime
+
+		// no record in ssbde.history, direct insert
+		if ssbde.history.Len() == 0 {
+			record := newStmtSummaryByDigestEvictedElement(eBeginTime, eEndTime)
+			record.addEvicted(evictedKey, evictedElement)
+			ssbde.history.PushFront(record)
+			continue
+		}
+
+		// prevent exceeding history size
+		for ssbde.history.Len() >= historySize && ssbde.history.Len() > 1 {
+			ssbde.history.Remove(ssbde.history.Front())
+		}
+
+		// look for matching history interval
+		// if there are no records in ssbde.history, following code will not be executed. which will probably lead to a bug.
```

Comment:
This comment seem to say there is a bug in some cases?

Reply:
Yes, if there are no records in history, the for loop will be jumped over, no matching or inserting will happen. This is a server bug.

So I constructed a direct insert to line 55~60 to fix it.